experimental/ComplexReflectionGroups/src/ComplexReflectionGroupType.jl

```diff
@@ -0,0 +1,680 @@
+                    return false
+                end
+            end
+        end
+    end
+    return true
+end
+
+function is_primitive(G::ComplexReflectionGroupType)
+    return !is_imprimitive(G)
+end
+
+function direct_product(X::Vector{ComplexReflectionGroupType})
+
+    type = reduce(vcat, [ G.type for G in X ]) #simply the flattened list of the types
+
+    return ComplexReflectionGroupType(type)
+
+end
+
+function direct_product(G1::ComplexReflectionGroupType, G2::ComplexReflectionGroupType)
+    return direct_product([G1,G2])
+end
+
+function num_reflections(G::ComplexReflectionGroupType)
+
+    N = ZZ(0)
+
+    # See Lehrer & Taylor (2009). TABLE D.3 (p 275)
+    excnum = [ 8, 16, 14, 22, 18, 30, 34, 46, 12, 18, 28, 34, 48, 78, 88, 118, 40, 70, 30, 15, 21, 24, 33, 45, 24, 40, 60, 60, 80, 45, 126, 36, 63, 120 ]
+
+    for t in G.type
+        if isa(t, Int)
+            i = t
+            N += ZZ(excnum[i-3]) #index shift by 3 because exceptionals start with G4
+        else
+            m = ZZ(t[1])
+            p = ZZ(t[2])
+            n = ZZ(t[3])
+            N += div(m*(n^2-n),2) + n*(div(m,p)-1) 
+            # Formula follows from Lehrer & Taylor (2009), Lemma 2.8.
+        end
+    end
+
+    return N
+end 
+
+function is_well_generated(G::ComplexReflectionGroupType)
+    
+    # Grabbed from Magma
+    excwell = [ 4, 5, 6, 8, 9, 10, 14, 16, 17, 18, 20, 21, 23, 24, 25, 26, 27, 28, 29, 30, 32, 33, 34, 35, 36, 37 ]
+    
+    for t in G.type
+        if isa(t, Int)
+            if !(t in excwell)
+                return false
+            end
+        else
+            # See Lehrer & Taylor (2009), Section 2.7 (p 35)
+            m = t[1]
+            p = t[2]
+            n = t[3]
+            if !(m == 1 || p == 1)
+                return false
+            end
+        end
+    end
+
+    return true
+end
+
+function degrees(G::ComplexReflectionGroupType)
+
+    degrees = Vector{Int}()
+
+    # Lehrer & Taylor (2009), Table D.3 (p 275)
+    excdegrees = 
+    [
+        [ 4, 6 ],
+        [ 6, 12 ],
+        [ 4, 12 ],
+        [ 12, 12 ],
+        [ 8, 12 ],
+        [ 8, 24 ],
+        [ 12, 24 ],
+        [ 24, 24 ],
+        [ 6, 8 ],
+        [ 8, 12 ],
+        [ 6, 24 ],
+        [ 12, 24 ],
+        [ 20, 30 ],
+        [ 20, 60 ],
+        [ 30, 60 ],
+        [ 60, 60 ],
+        [ 12, 30 ],
+        [ 12, 60 ],
+        [ 12, 20 ],
+        [ 2, 6, 10 ],
+        [ 4, 6, 14 ],
+        [ 6, 9, 12 ],
+        [ 6, 12, 18 ],
+        [ 6, 12, 30 ],
+        [ 2, 6, 8, 12 ],
+        [ 4, 8, 12, 20 ],
+        [ 2, 12, 20, 30 ],
+        [ 8, 12, 20, 24 ],
+        [ 12, 18, 24, 30 ],
+        [ 4, 6, 10, 12, 18 ],
+        [ 6, 12, 18, 24, 30, 42 ],
+        [ 2, 5, 6, 8, 9, 12 ],
+        [ 2, 6, 8, 10, 12, 14, 18 ],
+        [ 2, 8, 12, 14, 18, 20, 24, 30 ]
+    ]
+
+    for t in G.type
+        if isa(t, Int)
+            i = t
+            append!(degrees, convert(Vector{ZZRingElem}, excdegrees[i-3]) ) 
```

Comment:
This conversion is useless because `degrees` is initialized as a `Vector{Int}`, so everything gets converted back (also below).